File: 백준/Gold/20055. 컨베이어 벨트 위의 로봇/컨베이어 벨트 위의 로봇.py
import sys
from collections import deque
input = sys.stdin.readline

def solve():
    N, K = map(int, input().split())
    durability = deque(list(map(int, input().split()))) # 내구도
    robots = deque([False] * N) # 로봇 존재 여부 (위 N칸만 관리)

    step = 0
    while True:
        step += 1
        durability.rotate(1)
        robots.rotate(1)
        robots[N - 1] = False

        for i in range(N - 2, -1, -1):
            if robots[i] and not robots[i + 1] and durability[i + 1] >= 1:
                robots[i] = False
                robots[i + 1] = True
                durability[i + 1] -= 1
        robots[N - 1] = False

        if durability[0] > 0:
            robots[0] = True
            durability[0] -= 1
        
        if durability.count(0) >= K:
            print(step)
            break

if __name__ == "__main__":
    solve()

File: 백준/Gold/20055. 컨베이어 벨트 위의 로봇/컨베이어 벨트 위의 로봇.js
const fs = require("fs");
const input = fs.readFileSync(0).toString().trim().split("\n");

const [N, K] = input[0].split(" ").map(Number);
const durability = input[1].split(" ").map(Number);
const robots = Array(N - 1).fill(false);

let step = 0;
let zeroCount = 0; // 내구도가 0인 칸의 개수를 추적

for (let d of durability) {
  if (d === 0) zeroCount++;
}

while (true) {
  step++;

  durability.unshift(durability.pop());
  robots.pop();
  robots.unshift(false);

  robots[N - 1] = false;

  for (let i = N - 2; i >= 0; i--) {
    if (robots[i] && !robots[i + 1] && durability[i + 1] >= 1) {
      robots[i] = false;
      robots[i + 1] = true;
      durability[i + 1]--;

      if (durability[i + 1] === 0) zeroCount++;
    }
  }
  robots[N - 1] = false;

  if (durability[0] > 0) {
    robots[0] = true;
    durability[0]--;
    if (durability[0] === 0) zeroCount++;
  }

  if (zeroCount >= K) {
    console.log(step);
    break;
  }
}
